cpu/atxmega/periph/uart.c

```diff
@@ -0,0 +1,426 @@
+/*
+ * Copyright (C) 2021 Gerson Fernando Budke <nandojve@gmail.com>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_atxmega
+ * @ingroup     cpu_atxmega_periph
+ * @{
+ *
+ * @file
+ * @brief       Low-level UART driver implementation
+ *
+ * @author      Gerson Fernando Budke <nandojve@gmail.com>
+ *
+ *
+ * Supports runtime calculation of the BSEL and BSCALE register values.
+ * Supports reconfiguring UART after a clock change.
+ *
+ * @}
+ */
+#include <avr/io.h>
+
+#include <stdlib.h>
+#include <stdio.h>
+
+#include "board.h"
+#include "cpu.h"
+#include "sched.h"
+#include "thread.h"
+#include "periph/uart.h"
+#include "periph/gpio.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+/**
+ * @brief   UART Baudrate Tolerance
+ *
+ * The tolerance is expressed as baud rate percentage multiplied by 100.
+ * For bigger tolerances UART double frequency can be avoided, thus saving
+ * power.
+ *
+ * The default baud tolerance is 2%.
+ */
+#ifndef BAUD_TOL
+#define BAUD_TOL 2 * 100
+#endif
+
+/**
+ * @brief   Allocate memory to store the callback functions.
+ */
+static uart_isr_ctx_t isr_ctx[UART_NUMOF];
+
+/**
+ * @brief   Get the pointer to the base register of the given USART device
+ *
+ * @param[in] dev       USART device identifier
+ *
+ * @return              base register address
+ */
+static inline USART_t *dev(uart_t dev)
+{
+    return uart_config[dev].dev;
+}
+
+static inline int8_t _check_bsel(uint32_t *baud, uint32_t *calcbaud,
+                                 int16_t *precision)
+{
+    /* Avoid negative values and with precision of two positions
+     * after the decimal point for the deviation in percent
+     *
+     * result is the absolute deviation eg. 10 001
+     *
+     * MAX BAUD fper/2 = 16MHz a shift of 8 can be used to increase
+     * accuracy
+     */
+    uint16_t pre;
+
+    if (*baud < *calcbaud) {
+        pre = (uint16_t)((*calcbaud * 10000ULL) / *baud);
+    }
+    else {
+        pre = (uint16_t)((*baud * 10000ULL) / *calcbaud);
+    }
```

Comment:
```suggestion
    if (baud < calcbaud) {
        pre = (uint16_t)((calcbaud * 10000ULL) / baud);
    }
    else {
        pre = (uint16_t)((baud * 10000ULL) / calcbaud);
    }
```

I'd say it is more readable when passed by value.